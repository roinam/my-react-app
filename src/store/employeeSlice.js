import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { getAllEmployee } from '../service/employeeService';

// Delay Just for DEMO
const delay = async (ms) => {
  return new Promise((resolve) => 
      setTimeout(resolve, ms));
};

export const fetchAllEmployee = createAsyncThunk(
  'employee/fetchAllEmployee',
  async () => {
    const res = await getAllEmployee();//axios('http://localhost:8080/employee/findAll')
    //await delay(2000);
    const data = await res.data
    return data
  }
)

export const employeeSlice = createSlice({
  name: 'employee',
  initialState: {
    employeeList: [],
    theEmployee: {},
    isLoading: false,
    error: null
  },
  reducers: {
    storeEmployeeList: (state, action) => {
      return {...state, employeeList: action.payload};
    },
    storeTheEmployee: (state, action) => {
        return {...state, theEmployee: action.payload};
    },
    removeTheEmployee: (state, action) => {
        const newList = state.employeeList.filter(emp => emp.empId != action.payload);
        return {...state, employeeList: [...newList]};
    }
  },
  extraReducers: (builder) => {
    builder.addCase(fetchAllEmployee.pending, (state) => {
      state.isLoading = true
    })
    builder.addCase(fetchAllEmployee.fulfilled, (state, action) => {
      state.isLoading = false
      state.employeeList = action.payload
    })
    builder.addCase(fetchAllEmployee.rejected, (state, action) => {
      state.isLoading = false
      state.error = action.error.message
    })
  }
})

export const { storeEmployeeList, storeTheEmployee, removeTheEmployee } = employeeSlice.actions;