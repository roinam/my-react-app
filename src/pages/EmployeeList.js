import { useEffect } from "react";
import {useNavigate} from "react-router-dom";
import { deleteEmployee, getAllEmployee } from "../service/employeeService";
import { useDispatch, useSelector } from "react-redux"
import { fetchAllEmployee, removeTheEmployee, storeEmployeeList, storeTheEmployee } from "../store/employeeSlice";
import { PureComp } from "../component/PureComp";
import { ParentRenderProp } from "../component/rendererProp/ParentRenderProp";

const EmployeeList = (prop) => {
    const dispatch = useDispatch();
    const navigate = useNavigate();
    const allEmployeeList = useSelector(state => state.employee.employeeList);
    const isLoading = useSelector((state) => state.employee.isLoading)
    const errorInLoading = useSelector((state) => state.employee.error)

    useEffect(() => {
        const tryThunk = true;
        if(tryThunk) {
            dispatch(fetchAllEmployee());
        } else {
            getAllEmployee()
            .then(response => {
                if(response && response.data){
                    dispatch(storeEmployeeList(response.data));
                } else{
                    dispatch(storeEmployeeList([]));
                }
            })
            .catch(error => {
                dispatch(storeEmployeeList([]));
            });
        }
    }, []);

    // if(allEmployeeList.length==0)
    // throw new Promise((resolve) => setTimeout(resolve, 2000)); 

    const handleView = (employee) => {
        dispatch(storeTheEmployee({...employee}));
        navigate(`/view/${employee.empId}`);
    };

    const handleDelete = (id) => {
        deleteEmployee(id)
        .then(response => {
            if(response && response.data){
                dispatch(removeTheEmployee(id));
            }
        })
        .catch(() => {});
    };

    const handleAdd = () => {
        navigate("/add");
    }

    if(isLoading) {
        return <>
            <div className="container">
                Loading.....
            </div>
        </>;
    }

    if(errorInLoading) {
        return <>
            <div className="container">
                Sorry! Unable to load data
                <div className="mt-3" style={{color:"red"}}>{errorInLoading}</div>
            </div>
        </>;
    }

    return <>
    <PureComp name={"Testing"}></PureComp>
    <ParentRenderProp></ParentRenderProp>
    
    <div className="d-flex flex-row mb-3">
        <button type="button" className="btn btn-primary" onClick={handleAdd}>
            <i className="fa-solid fa-plus"></i>
            Add New Employee
        </button>

        {/* <button type="button" className="btn btn-primary" style={{marginLeft:"3px"}}
        onClick={() => {throw new Error('Crashed!!!!');}}>
            <i className="fa-solid fa-plus"></i>
            Show Error
        </button> */}
    </div>

        <table className="table table-striped">
        <thead className="table-dark">
            <tr>
                <th scope="col">Emp#</th>
                <th scope="col">Name</th>
                <th scope="col">Designation</th>
                <th scope="col">Department</th>
                <th scope="col">Salary</th>
                <th scope="col" style={{width:"80px"}}></th>
                <th scope="col" style={{width:"80px"}}></th>
            </tr>
        </thead>
        <tbody>
            {allEmployeeList && allEmployeeList.map(employee => 
                <tr key={employee.empId}>
                    <th scope="row">{employee.empId}</th>
                    <td>{employee.empName}</td>
                    <td>{employee.designation}</td>
                    <td>{employee.department}</td>
                    <td>{employee.salary}</td>
                    <td>
                    <button type="button" className="btn btn-outline-info" onClick={() => handleView(employee)}>View</button>
                    </td>
                    <td>
                    <button type="button" className="btn btn-outline-danger" onClick={() => handleDelete(employee.empId)}>Delete</button>
                    </td>
                </tr>
            )}
        </tbody>
        </table>
    </>;
};

export default EmployeeList;