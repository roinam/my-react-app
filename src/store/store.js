import { configureStore } from "@reduxjs/toolkit";
import { employeeSlice } from "./employeeSlice";

export const store = configureStore({
    reducer: {employee: employeeSlice.reducer}
});