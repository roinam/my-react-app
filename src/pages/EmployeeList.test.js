import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from "react-redux";
import { EmployeeList } from "./EmployeeList";
import { store } from '../store/store';
import renderer from 'react-test-renderer';

  jest.mock('react-router-dom', () => {
      // Require the original module to not be mocked...
      const originalModule = jest.requireActual('react-router-dom');
    
      return {
        __esModule: true,
        ...originalModule,
        useParams: jest.fn(),
        useHistory: jest.fn(),
        useNavigate: jest.fn()
      };
  });

  const data = {data: [
    {empId: 2, empName:"Test User 2", designation:"test designation", department:"test dept", salary:"20000"}
  ]};
  jest.mock('axios', () => {
    return {
      create: jest.fn((parm) => {
        return {
          get: () => Promise.resolve(data)
        }
      })
    };
  });

beforeAll(() => {
    
});

describe('EmployeeList Test Suit', () => {
  test('should create component with axios', async () => {
    render(<Provider store={store}><EmployeeList></EmployeeList></Provider>);

    await waitFor(() => {
      const theElement = screen.getByText(/Test User 2/i);
      expect(theElement).toBeInTheDocument();
    },
    {
      timeout: 3000
    });    
  });

  test('snapshot renders correctly', () => {
    const tree = renderer
      .create(<Provider store={store}><EmployeeList></EmployeeList></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});