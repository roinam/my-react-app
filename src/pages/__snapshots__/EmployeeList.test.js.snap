// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeList Test Suit snapshot renders correctly 1`] = `
Array [
  <div
    className="d-flex flex-row mb-3"
  >
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa-solid fa-plus"
      />
      Add New Employee
    </button>
  </div>,
  <table
    className="table table-striped"
  >
    <thead
      className="table-dark"
    >
      <tr>
        <th
          scope="col"
        >
          Emp#
        </th>
        <th
          scope="col"
        >
          Name
        </th>
        <th
          scope="col"
        >
          Designation
        </th>
        <th
          scope="col"
        >
          Department
        </th>
        <th
          scope="col"
        >
          Salary
        </th>
        <th
          scope="col"
          style={
            Object {
              "width": "80px",
            }
          }
        />
        <th
          scope="col"
          style={
            Object {
              "width": "80px",
            }
          }
        />
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          2
        </th>
        <td>
          Test User 2
        </td>
        <td>
          test designation
        </td>
        <td>
          test dept
        </td>
        <td>
          20000
        </td>
        <td>
          <button
            className="btn btn-outline-info"
            onClick={[Function]}
            type="button"
          >
            View
          </button>
        </td>
        <td>
          <button
            className="btn btn-outline-danger"
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>,
]
`;
